kernel/src/process/wait.rs

```diff
@@ -96,16 +101,84 @@ pub fn wait_child_exit(
     Ok(zombie_child)
 }
 
+pub enum WaitStatus {
+    Process(Arc<Process>),
+    Thread(Arc<Thread>, ThreadWaitStatus),
+}
+
+impl WaitStatus {
+    pub fn id(&self) -> u32 {
+        match self {
+            WaitStatus::Process(process) => process.pid(),
+            WaitStatus::Thread(thread, _) => thread.as_posix_thread().unwrap().tid(),
+        }
+    }
+
+    pub fn exit_status(&self) -> u32 {
+        match self {
+            WaitStatus::Process(process) => process.status().exit_code(),
+            WaitStatus::Thread(_, thread_wait_status) => thread_wait_status.as_u32(),
+        }
+    }
+
+    pub fn prof_clock(&self) -> &Arc<ProfClock> {
+        match self {
+            WaitStatus::Process(process) => process.prof_clock(),
+            WaitStatus::Thread(thread, _) => thread.as_posix_thread().unwrap().prof_clock(),
+        }
+    }
+}
+
+fn wait_child_zombie(
+    unwaited_children: &[Arc<Process>],
+    wait_options: WaitOptions,
+    children_lock: &mut BTreeMap<Pid, Arc<Process>>,
+) -> Option<WaitStatus> {
+    let zombie_child = unwaited_children
+        .iter()
+        .find(|child| child.status().is_zombie())?;
+
+    let wait_status = WaitStatus::Process(zombie_child.clone());
+    if wait_options.contains(WaitOptions::WNOWAIT) {
+        Some(wait_status)
+    } else {
+        let zombie_pid = zombie_child.pid();
+        reap_zombie_child(zombie_pid, children_lock);
+        Some(wait_status)
+    }
+}
+
+fn wait_thread(
```

Comment:
I don't believe that any of `SIGSTOP`, `SIGCONT`, `WSTOPPED`, and `WCONTINUED` are related to threads. They affect the entire process.

Similarly, you cannot SIGKILL a single thread. It will kill the entire process.